File: medium/64 (Minimum path sum)/solution.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:

        dp = []
        for i in range(0,len(grid)):
            dp.append([])
            for y in range(0,len(grid[i])):
                z = 0
                if i == 0 and y == 0:
                    z = grid[i][y]
                elif i == 0:
                    z = grid[i][y] + dp[i][y - 1]
                elif y == 0:
                    z = grid[i][y] + dp[i - 1][y]
                else:
                    z = grid[i][y] + min(dp[i-1][y],dp[i][y - 1])
                dp[i].append(z)

        return dp[len(grid) - 1][len(grid[0]) - 1]

File: medium/64 (Minimum path sum)/solution.java
class Solution {
    public int minPathSum(int[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        int [][]dp = new int[n][m];
        for(int i = 0; i < n; i++){
            for(int y = 0; y < m; y++){
                if(i == 0 && y == 0) dp[i][y] = grid[i][y];
                else if(i == 0) dp[i][y] = dp[i][y - 1] + grid[i][y];
                else if(y == 0) dp[i][y] = dp[i - 1][y] + grid[i][y];
                else dp[i][y] = grid[i][y] + Math.min(dp[i-1][y],dp[i][y - 1]);
            }
        }
        return dp[n - 1][m - 1];
    }
}